core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -314,31 +319,51 @@ class ReassignPartitionsCommand(zkUtils: ZkUtils, partitions: Map[TopicAndPartit
+    val topicsForMove = partitions.map { case (tp, _) => tp.topic }.toSeq
+    val current = zkUtils.getReplicaAssignmentForTopics(topicsForMove)
+    assignThrottledReplicas(current, partitions)
+  }
+
+  private[admin] def assignThrottledReplicas(allExisting: Map[TopicAndPartition, Seq[Int]], allProposed: Map[TopicAndPartition, Seq[Int]], admin: AdminUtilities = AdminUtils): Unit = {
+    for (topic <- allProposed.keySet.map(tp => tp.topic).toSeq.distinct) {
+      val (existing, proposed) = filterBy(topic, allExisting, allProposed)
+
+      //Apply the leader throttle to all replicas that exist before the re-balance.
+      val leader = format(preRebalanceReplicaForMovingPartitions(existing, proposed))
+
+      //Apply a follower throttle to all "move destinations".
+      val follower = format(postRebalanceReplicasThatMoved(existing, proposed))
+
+      admin.changeTopicConfig(zkUtils, topic, wrap((LeaderThrottledReplicasListProp, leader), (FollowerThrottledReplicasListProp, follower)))
+      info(s"Updated leader-throttled replicas for topic $topic with: $leader")
+      info(s"Updated follower-throttled replicas for topic $topic with: $follower")
+    }
+  }
+
+  private def postRebalanceReplicasThatMoved(existing: Map[TopicAndPartition, Seq[Int]], proposed: Map[TopicAndPartition, Seq[Int]]): Map[TopicAndPartition, Seq[Int]] = {
+    //For each partition in the proposed list, filter out any replicas that exist now (i.e. not moving)
+    existing.map { case (tp, current) =>
+      (tp -> proposed.get(tp).get.filterNot(current.toSet))
+    }
   }
 
-  def replicaMoves(existing: Map[TopicAndPartition, Seq[Int]], proposed: Map[TopicAndPartition, Seq[Int]]): Map[String, String] = {
-    //Find the replicas that have moved
-    val movesByPartition = existing.map { case (topicAndPartition, existingReplicas) =>
-      val before = existingReplicas.toSet
-      val after = proposed.get(topicAndPartition).get.toSet
-      val moving = (after.filterNot(before) ++ before.filterNot(after)).toSeq.distinct.sorted
-      val formatted = moving.map { brokerId => s"${topicAndPartition.partition}:$brokerId" }
-      (topicAndPartition, formatted)
+  private def preRebalanceReplicaForMovingPartitions(existing: Map[TopicAndPartition, Seq[Int]], proposed: Map[TopicAndPartition, Seq[Int]]): Map[TopicAndPartition, Seq[Int]] = {
+    //Throttle all existing replicas (as any one might be a leader). So just filter out those which aren't moving
+    existing.filter { case (tp, current) =>
+      proposed.get(tp).get.filterNot(current.toSet).size > 0
     }
-    //Group by topic
-    val movesByTopic = movesByPartition.groupBy(_._1.topic)
-      .map { case (topic, reps) => (topic, reps.values.flatMap(rep => rep)) }
+  }
+
+  def format(moves: Map[TopicAndPartition, Seq[Int]]): String = {
+    val formatted = moves.map { case (tp, moves) => moves.map { replicaId => s"${tp.partition}:${replicaId}" } }.flatMap(x => x).mkString(",")
```

Comment:
I'd write this like:

```scala
    val formatted = moves.flatMap { case (tp, moves) =>
      moves.map(replicaId => s"${tp.partition}:${replicaId}")
    }.mkString(",")
```

It's good to break down things instead of having it all in one line. Also, I removed what looked like an unnecessary `flatMap`.